docs(themes): Update working-with-themes.md (#12295)
unoplatform/uno: doc/articles/features/working-with-themes.md

uid: Uno.Features.Themes
---

# Supporting Dark and High Contrast themes

Uno Platform offers fine-grained customization of typography, corner radius, and styling of UI elements to convey a brand identity in your application. However, platforms like iOS, Android, and web browsers also offer a system-wide color mode that users can change. For instance, Windows uses **Light mode** by default, but many users prefer **Dark mode** which is intended to be easier on the eyes in low-light conditions. Many platforms also offer a **High Contrast mode** to make things easier to see.

> These themes affect both the `Background` and `Foreground` colors to accomodate user preferences. All of the color modes mentioned above are available for use in your app. This guide will detail how to change the system theme setting, make your app recieve change notifications for it, and react to those changes at runtime.

Reviewer: ```suggestion
These themes affect both the `Background` and `Foreground` colors to accommodate user preferences. All of the color modes mentioned above are available for use in your app. This guide will detail how to change the system theme setting, make your app receive change notifications for it, and react to those changes at runtime.
```

## Enable dark mode

> As in WinUI, the possible values `Light`, `Dark`, and `HighContrast` correspond to a value users can select in the settings of their respective platform. 

Reviewer: ```suggestion
As in WinUI, the possible values `Light`, `Dark`, and `HighContrast` correspond to a value users can select in the settings of their respective platforms. 
```

> High Contrast mode is often available to enable seperately, from a dedicated _Accessibility_ page. For the purpose of this documentation, we will assume the user wants to use Dark mode.

Reviewer: ```suggestion
High Contrast mode is often available to enable separately, from a dedicated _Accessibility_ page. For the purpose of this documentation, we will assume the user wants to use Dark mode.
```

# [**iOS**](#tab/ios)
Devices running iOS or iPadOS can enable Dark mode from Control Center or Settings. See [this guide](https://support.apple.com/HT210332) for more information.

# [**Android**](#tab/android)
Android devices can enable Dark mode from Settings. See [this guide](https://support.google.com/android/answer/9730472) for more information.

# [**Browser**](#tab/browser)
The steps to enable Dark mode in a browser vary by browser. See the following guides for more information:

- [Chrome](https://support.google.com/chrome/answer/9275525)
- [Edge](https://support.microsoft.com/microsoft-edge/use-the-dark-theme-in-microsoft-edge-9b74617b-f542-77ed-033b-1a5cfb17a2df)
- [Firefox](http://mzl.la/1BAQGDX)

---

## React to OS theme changes

When you change the theme mode on your device, the system will send a notification to your app. The colors of UI elements in your app will automatically switch over as long as you do _not_ specify a theme in any of the following places:

* `App` constructor
* `App.xaml` 
* `AppResources.xaml`
* The `RequestedTheme` property of a parent `FrameworkElement`

> However, your app can still react manually to changes in the theme mode. To do so, you can use the [Uno.CommunityToolkit.WinUI.UI](https://www.nuget.org/packages/Uno.CommunityToolkit.WinUI.UI) NuGet package. This package is not required to change the app color theme, but it contains a `ThemeListener` class that can be used to listen for OS theme changes. To actually change the app color theme at runtime, you need to install the [Uno.Toolkit.UI](https://www.nuget.org/packages/Uno.Toolkit.UI) NuGet package which contains a `SystemThemeHelper` class.

Author: @kazo0 I agree in principle that requiring a third party library for this rather fundamental feature doesn't have the best look. What should the next steps be, given that the doc now is functionally unhelpful and very out of date? I can open an issue to track exposing the existing code in Uno.UWP from the Toolkit SystemThemeHelper.

Would you consider it a roadblock for this pull request? I feel like getting _something_  out that is improved should be the priority until we expose a proper theme listener in the toolkit.

> However, your app can still react manually to changes in the theme mode. To do so, you can use the [Uno.CommunityToolkit.WinUI.UI](https://www.nuget.org/packages/Uno.CommunityToolkit.WinUI.UI) NuGet package. This package is not required to change the app color theme, but it contains a `ThemeListener` class that can be used to listen for OS theme changes. To actually change the app color theme at runtime, you need to install the [Uno.Toolkit.UI](https://www.nuget.org/packages/Uno.Toolkit.UI) NuGet package which contains a `SystemThemeHelper` class.

Reviewer 2: We can publish the doc like this for the current state. 
Just add a bullet point to update the doc as part of that issue.

> However, your app can still react manually to changes in the theme mode. To do so, you can use the [Uno.CommunityToolkit.WinUI.UI](https://www.nuget.org/packages/Uno.CommunityToolkit.WinUI.UI) NuGet package. This package is not required to change the app color theme, but it contains a `ThemeListener` class that can be used to listen for OS theme changes. To actually change the app color theme at runtime, you need to install the [Uno.Toolkit.UI](https://www.nuget.org/packages/Uno.Toolkit.UI) NuGet package which contains a `SystemThemeHelper` class.

Author: @Xiaoy312 Is this PR ready to approve, then?

Below is an example of how to use these classes to change the app theme at runtime based on a raised event:

```csharp
using CommunityToolkit.WinUI.UI.Helpers;
using Uno.Toolkit.UI;

namespace MyApp.UI.Views;

public class MainPage : Page
{
    public MainPage()
    {
        this.InitializeComponent();

        ThemeListener.Current.ThemeChanged += OnThemeChanged;
    }

    private void OnThemeChanged(ThemeListener sender)
    {
        bool isDarkMode = sender.CurrentTheme == ApplicationTheme.Dark;

        SystemThemeHelper.SetApplicationTheme(darkMode: isDarkMode);
    }
}
```

## Change the app theme at runtime

> The `SetApplicationTheme` method in this helper class can also be used to support an in-app toggle for dark mode. For example, you could add a toggle button to your app's settings page that allows the user to switch between light and dark mode. The following code snippet shows how to implement this:

Reviewer: ```suggestion
The `SetApplicationTheme` method in this helper class can also be used to support an in-app toggle for dark mode. For example, you could add a toggle button to your app's settings page that allows the user to switch between light and dark modes. The following code snippet shows how to implement this:
```

> The `SetApplicationTheme` method in this helper class can also be used to support an in-app toggle for dark mode. For example, you could add a toggle button to your app's settings page that allows the user to switch between light and dark modes. The following code snippet shows how to implement this:

Reviewer 2: be specific in what you refers to as: "in this helper class"
```suggestion
The [`SystemThemeHelper.SetApplicationTheme`](https://github.com/unoplatform/uno.toolkit.ui/blob/main/src/Uno.Toolkit.UI/Helpers/SystemThemeHelper.cs) method from `Uno.Toolkit.UI` can also be used to support an in-app toggle for dark mode. For example, you could add a toggle button to your app's settings page that allows the user to switch between light and dark modes. The following code snippet shows how to implement this:
```

```csharp

> using CommunityToolkit.WinUI.UI.Helpers;

Reviewer 2: doesnt seem to be needed for the part of code shown:
```suggestion
using CommunityToolkit.WinUI.UI.Helpers;
```

using Uno.Toolkit.UI;

namespace MyApp.UI.Views;

public class SettingsPage : Page
{
    public SettingsPage()
    {
        this.InitializeComponent();

        this.Loaded += OnLoaded;
    }

    private void OnLoaded(object sender, RoutedEventArgs e)
    {
        this.DarkModeToggle.IsOn = SystemThemeHelper.IsDarkModeEnabled;
    }

    private void OnDarkModeToggleToggled(object sender, RoutedEventArgs e)
    {
        SystemThemeHelper.SetApplicationTheme(darkMode: this.DarkModeToggle.IsOn);
    }
}
```

## Change the app theme at startup

> Another method to change the app theme is to adjust it immediately upon startup. The following must only be called in the `App` class constructor

Reviewer: ```suggestion
Another method to change the app theme is to adjust it immediately upon startup. The following must only be called in the `App` class constructor.
```

```csharp

> Uno.UI.ApplicationHelper.RequestedCustomTheme = "Dark";

Reviewer 2: you should just include the full code (containing method signature, uno vs windows code), so it is clear at a glance:
```suggestion
public App()
{
	this.InitializeComponent();

#if HAS_UNO
	Uno.UI.ApplicationHelper.RequestedCustomTheme = nameof(ApplicationTheme.Dark);
#else
	this.RequestedTheme = ApplicationTheme.Dark;
#endif
}
```
